### DISCUSSION TITLE
Reasons for the PNPM server (package manager architecture question)?

### DISCUSSION THREAD

**Discussion by lancejpollard** [pnpm/pnpm#8136] [Q&A]
I am looking through the PNPM code for inspiration to build a package manager for a custom language, and am wondering what the purpose of the PNPM server is?

ChatGPT is saying that there are three scenarios to use it:

1. Solo developer loading packages from a local server cache.
2. Company/team developers loading packages from local network cache.
3. CI/CD, where you need faster package installation from a cache (but how does that work, in reality?).

It is saying that it is faster in both scenarios because it's using the cached package instead of fetching from NPM, but my initial questions are:

1. Isn't it still fetching from the internet anyways, so it still has latency? Maybe I guess npmjs.org latency is worse than "my own server" hosted on AWS or Vercel or something? Or how are you thinking about this / approaching this?
2. Aren't the packages only installed once in the PNPM store locally on the file system, so it won't fetch packages more than once? ChatGPT was saying that was a benefit of the server, but I thought the file system cached everything already. So I can only think that perhaps a local server storing everything in memory would be faster than the file system, but maybe I misreading some stuff.

Basically my questions are:

1. What are the main reasons for the PNPM server? How does it improve performance exactly, can you dig a little beneath the surface?
2. When you choose to use it, what benefit would it give you, what are the tradeoffs?
3. What does the normal "default" mode of the PNPM file store do differently than the server?
4. If you use PNPM server on CI/CD, how does that work in theory?

Thank you so much for the help!
